rotkehlchen/tests/unit/decoders/test_cowswap.py

```diff
@@ -0,0 +1,366 @@
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [[
+    '0x0D2f07876685bEcd81DDa1C897f2D6Cacc733fc1',
+    '0x34938Bd809BDf57178df6DF523759B4083A29190',
+]])
+def test_swap_token_to_token_and_token_to_eth(database, ethereum_inquirer, ethereum_accounts):
+    tx_hex = deserialize_evm_tx_hash('0xd4d16ea74bbf806715f5f0e799fd5e8befbf369a9e5461fa9c0ed88d72bd06e4')  # noqa: E501
+    evmhash = deserialize_evm_tx_hash(tx_hex)
+    user_address_1, user_address_2 = ethereum_accounts
+    events, _ = get_decoded_events_of_transaction(
+        evm_inquirer=ethereum_inquirer,
+        database=database,
+        tx_hash=tx_hex,
+    )
+    expected_events = [
+        HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=1,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            asset=Asset('eip155:1/erc20:0xe9B076B476D8865cDF79D1Cf7DF420EE397a7f75'),
+            balance=Balance(amount=FVal('16000')),
+            location_label=user_address_1,
+            notes='Swap 16000 FUND in cowswap',
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=2,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=A_WETH,
+            balance=Balance(amount=FVal('4.870994011222719015')),
+            location_label=user_address_1,
+            notes='Receive 4.870994011222719015 WETH as the result of a swap in cowswap',  # noqa: E501
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=3,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            asset=A_USDT,
+            balance=Balance(amount=FVal('99.99')),
+            location_label=user_address_2,
+            notes='Swap 99.99 USDT in cowswap',
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=4,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=A_ETH,
+            balance=Balance(amount=FVal('0.053419767450716028')),
+            location_label=user_address_2,
+            notes='Receive 0.053419767450716028 ETH as the result of a swap in cowswap',  # noqa: E501
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=5,
```

Comment:
This is what I meant. you are putting the approvals at the end with what you are doing.